### PR TITLE
[muon] add Gram Newton-Schulz algorithm + batched standard NS for square groups (#187132)

### PR DESCRIPTION
**Pull request by tangyuhao**
Summary:

## Context

Diff 2 of the upstream port of the Gram Muon stack. Diff 1 added the multi-tensor (`foreach`) scaffolding for the standard NS path with no algorithmic changes. This diff introduces the alternate orthogonalization algorithm itself plus the bmm-batched standard NS fast path for same-shape square groups (the perf piece from `D105093722`).

The selector is now an enum (`NewtonSchulzAlgorithm`) rather than a `use_gram_newton_schulz: bool`, addressing the "boolean trap" reviewer comment on the abandoned `D104311008`.

## This diff

1. **New public API**:
   - `NewtonSchulzAlgorithm` -- a `str`-backed `Enum` with values `STANDARD` and `GRAM`. Registered with `torch.serialization.add_safe_globals` so state dicts containing it deserialize cleanly under `torch.load(weights_only=True)`.
   - `GramNewtonSchulzConfig` -- `TypedDict` with optional `gram_ns_coefficients` (per-iteration `(a, b, c)` triples) and `gram_ns_reset_iterations`.
   - `Muon.__init__` gains `ns_algorithm: NewtonSchulzAlgorithm = STANDARD` and `gram_newton_schulz_config: GramNewtonSchulzConfig | None = None`. The config is rejected (`ValueError`) when `ns_algorithm` is not `GRAM` so callers cannot silently configure gram-only knobs.

2. **Algorithms** (lifted from the abandoned `D104311008` plus the batched-square path from `D105093722`):
   - `_gram_newton_schulz`, `_gram_newton_schulz_batched` -- Gram-space iteration (`R = X @ X^T`, accumulate `Q`, apply `X = Q @ X` at the end). Batched variant operates on `(batch, M, M)` via `bmm`/`baddbmm`.
   - `_zeropower_via_gram_newtonschulz`, `_zeropower_via_gram_newtonschulz_batched` -- the wrappers (normalize, half-precision cast, transpose to land on `M <= N`). Single-tensor variant falls back to `_zeropower_via_newtonschulz` for square shapes.
   - `_zeropower_via_newtonschulz_batched` -- batched standard NS for same-shape square groups via `bmm`/`baddbmm` (ported from `D105093722`).

3. **Dispatch**: new `_ns_eager` routes one same-shape group by `(ns_algorithm, is_square, len)`:
   - STANDARD, square, len > 1: batched `bmm/baddbmm`.
   - STANDARD, otherwise: per-tensor standard NS.
   - GRAM, rectangular, len > 1: batched Gram NS.
   - GRAM, rectangular, len == 1: single Gram NS.
   - GRAM, square: fall back to the standard routing (Gram has no advantage on square shapes; the single-tensor wrapper already falls back internally).
   `_run_ns` shape-groups the updates and calls `_ns_eager` per group. `_single_tensor_muon` calls the per-tensor gram/standard routine directly; `_multi_tensor_muon`'s Phase 2 calls `_run_ns`.

4. **State-dict BC**: `Muon.__setstate__` backfills defaults for `ns_algorithm`, `gram_ns_coefficients`, and `gram_ns_reset_iterations` so state dicts saved by older versions of Muon load cleanly. It also normalizes `ns_algorithm` from a plain string back to the enum (since the enum subclasses `str`, JSON / external round-trips can leave it as a raw string).

5. **Reviewer nits applied to the new code**: `strict=True` on every `zip()` (the only new `zip` is in `_run_ns`); the `.half()` placement inside the gram normalizer is folded into a single expression (`X = (X / (X.norm() + eps)).half()`) to avoid an intermediate fp32 allocation.

6. **Numerics caveat**: Diff 1's `_multi_tensor_muon` was bit-identical to single-tensor. Diff 2 changes that for one case: when `foreach=True` and STANDARD with same-shape square groups of `len > 1`, those groups now go through batched `bmm/baddbmm` instead of per-tensor `mm/addmm`. Reduction order across the batch dim differs, so the result is within float-precision tolerance (`atol=1e-5, rtol=1e-5` in practice) but not bit-equivalent. This is documented in both the class docstring and the `foreach` argument doc.

7. **Test coverage**:
   - `optim_inputs_func_muon` is unchanged in shape (the gram input was prototyped but ultimately removed because gram NS on the test's rank-1 `(10, 5)` gradient does not converge reliably with default coefficients in 20 steps; gram is exercised by dedicated tests instead).
   - `optim_error_inputs_func_muon` gains two error inputs: passing `gram_newton_schulz_config` with `ns_algorithm=STANDARD` must `ValueError`; passing an unknown key in the config must `ValueError`.
   - New `TestMuon` test class in `test_optim.py` (no separate `test_muon.py` -- per reviewer feedback we keep Muon-only behavior in `test_optim.py`):
     - `test_gram_vs_standard_differ_on_rectangular` -- different update directions on `(8, 16)`.
     - `test_gram_square_falls_back_to_standard` -- `atol=0, rtol=0` parity on `(8, 8)` between the two algorithms (square fallback).
     - `test_batched_standard_square_matches_per_tensor` -- batched `bmm` (foreach=True) within tolerance of per-tensor for three `(8, 8)` params.
     - `test_gram_converges_on_toy_regression` -- 200 steps drop loss by >= 90% on a 2-layer linear regression with rectangular weights.
     - `test_ns_algorithm_accepts_str` -- the str-backed enum lets callers pass `"gram"` directly.
     - `test_state_dict_round_trip_normalizes_ns_algorithm` -- a stale state dict (string `"gram"`, no gram keys) reconstitutes correctly via `__setstate__` and `step()` succeeds.

Authored by Claude.

Test Plan:
## Lint

```
arc lint -a fbcode/caffe2/torch/optim/_muon.py fbcode/caffe2/torch/testing/_internal/common_optimizers.py fbcode/caffe2/test/test_optim.py xplat/caffe2/torch/optim/_muon.py xplat/caffe2/torch/testing/_internal/common_optimizers.py xplat/caffe2/test/test_optim.py
ok No lint issues.
```

## Muon-related parametrized suite

```
buck2 test fbcode//caffe2/test:optim -- --regex 'Muon|TestMuon'
Tests finished: Pass 51. Fail 0. Timeout 0. Fatal 0. Skip 4. Omit 0. Infra Failure 0. Build failure 0
```

Same 4 multi-GPU/CUDA skips as Diff 1. Net +6 tests vs Diff 1 -- the `TestMuon` class added in this diff:

- `test_gram_vs_standard_differ_on_rectangular`
- `test_gram_square_falls_back_to_standard`
- `test_batched_standard_square_matches_per_tensor`
- `test_gram_converges_on_toy_regression`
- `test_ns_algorithm_accepts_str`
- `test_state_dict_round_trip_normalizes_ns_algorithm`

## Targeted re-run of just the new tests

```
buck2 test fbcode//caffe2/test:optim -- --regex 'test_gram_vs_standard_differ_on_rectangular|test_gram_square_falls_back_to_standard|test_batched_standard_square_matches_per_tensor|test_gram_converges_on_toy_regression|test_ns_algorithm_accepts_str|test_state_dict_round_trip_normalizes_ns_algorithm'
Tests finished: Pass 6. Fail 0. Timeout 0. Fatal 0. Skip 0. Omit 0. Infra Failure 0. Build failure 0
```

## No-regression sanity on the broader foreach parity matrix

```
buck2 test fbcode//caffe2/test:optim -- --regex 'test_foreach_matches_forloop'
Tests finished: Pass 13. Fail 0. Timeout 0. Fatal 0. Skip 0. Omit 0. Infra Failure 0. Build failure 0
```

(Re-validated from Diff 1; the Diff 2 batched-bmm path for STANDARD is only engaged when there are same-shape square groups with len > 1, which the standard `_test_derived_optimizers` setup does not produce -- so the foreach-parity test stays bit-identical.)

Differential Revision: D105621586

### CODE REVIEW

File: torch/optim/_muon.py [hunk 1]
```diff
@@ -70,6 +109,186 @@ def _zeropower_via_newtonschulz(
...
+    """Batched standard NS for a list of same-shape square 2D gradients.
+
+    Stacks gradients into a 3D batch and runs the quintic NS iteration with
+    bmm/baddbmm so same-shape squares fuse into one batched matmul per op
+    instead of sequential per-tensor mm/addmm. Caller guarantees all entries
+    are 2D and square with identical shape.
+    """
+    a, b, c = ns_coefficients
+    X = torch.stack([g.bfloat16() for g in grads])  # (batch, M, M)
+    X = X / X.norm(dim=(-2, -1), keepdim=True).clamp(min=eps)
```
Comment (Skylion007) [inline torch/optim/_muon.py:127]: /= should be better can be inplace can't it?

File: torch/optim/_muon.py [hunk 2]
```diff
@@ -70,6 +109,186 @@ def _zeropower_via_newtonschulz(
...
+    for _ in range(ns_steps):
+        gram = X @ X.mT
+        gram_update = torch.baddbmm(gram, gram, gram, beta=b, alpha=c)
+        X = torch.baddbmm(X, gram_update, X, beta=a)
```
Comment (Skylion007) [inline torch/optim/_muon.py:131]: Likewise does it support it being in place here?
